examples/small-coalton-programs/src/brainfold.lisp

```diff
@@ -0,0 +1,341 @@
+
+  (define-type Cmd
+    BFRight
+    BFLeft
+    BFPlus
+    BFMinus
+    BFPrint
+    BFInput
+    (BFLoop (List Cmd)))
+
+  (declare split-string (String -> List String))
+  (define (split-string s)
+    "Takes a string, returns a list of characters."
+    (if (arith:zero? (str:length s))
+        nil
+        (cons (substring s 0 1)
+              (split-string (substring s 1 (str:length s))))))
+
+  (declare index-after-loop ((List String) -> UFix))
+  (define (index-after-loop cmd-strings)
+    "Takes a list of bf strings, returns the count after the end of the
+corresponding loop."
+    (let ((until-end-loop-count
+            (fn (cmd-strings depth cmd-count)
+              (match cmd-strings
+                ((Nil) (error "Incomplete Loop"))
+                ((Cons "]" _)
+                 (if (arith:zero? depth)
+                     cmd-count
+                     (until-end-loop-count (list:cdr cmd-strings)
+                                           (1- depth)
+                                           (1+ cmd-count))))
+                ((Cons "[" _)
+                 (until-end-loop-count (list:cdr cmd-strings)
+                                       (1+ depth)
+                                       (1+ cmd-count)))
+                (_ (until-end-loop-count (list:cdr cmd-strings)
+                                         depth
+                                         (1+ cmd-count)))))))
+      (until-end-loop-count cmd-strings 0 2)))
+
+  (declare parse (String -> (List Cmd)))
+  (define (parse input-string)
+    "Parses Brainfold input into a lexical tree."
+    (let ((match-cmds
+            (fn (str-list)
+              (match str-list
+                ((Nil) Nil)
+                ((Cons ">" _)
+                 (Cons BFRight (match-cmds (list:cdr str-list))))
+                ((Cons "<" _)
+                 (Cons BFLeft (match-cmds (list:cdr str-list))))
+                ((Cons "+" _)
+                 (Cons BFPlus (match-cmds (list:cdr str-list))))
+                ((Cons "-" _)
+                 (Cons BFMinus (match-cmds (list:cdr str-list))))
+                ((Cons "." _)
+                 (Cons BFPrint (match-cmds (list:cdr str-list))))
+                ((Cons "," _)
+                 (Cons BFInput (match-cmds (list:cdr str-list))))
+                ((Cons "[" _)
+                 (Cons (BFLoop (match-cmds (list:cdr str-list)))
+                       (match-cmds
+                        (list:nth-cdr
+                         (index-after-loop (list:cdr str-list))
+                         str-list))))
+                ((Cons "]" _) nil)
+                (_ (match-cmds (list:cdr str-list)))
+                ))))
+      (match-cmds (split-string input-string)))))
+
+
+;;;
+;;; Evaluation/Compilation
+;;;
+
+(coalton-toplevel
+
+  (define (exec cmd)
+    "Executes a given bf command string."
+    (match cmd
+      ((BFRight) (move-right))
+      ((BFLeft) (move-left))
+      ((BFPlus) (fromint (incr)))
```

Comment:
I think elsewhere we should specify the type of the array so we don't need to `fromInt` everything everywhere.